### ISSUE TITLE
EXPLAIN fails for parameterized router plans

### ISSUE THREAD

**Issue by yglh-gq** [citusdata/citus#2009]
# Repro Steps

Start a Citus cluster in the background and connect to the coordinator…

```sh
mkdir citus_2009 && cd citus_2009

curl -sLO https://raw.githubusercontent.com/citusdata/docker/v7.3.0/docker-compose.yml

docker-compose up -d
# Creating citus_master ... done
# Creating citus_manager ... done
# Creating citus2009_worker_1  ... done

psql -p 5432 -h localhost -U postgres -X
# psql (10.3 (Homebrew petere/postgresql))
# ...
```

Connect to the coordinator and run the following…

```sql
CREATE TABLE simple (id integer, name text);
# CREATE TABLE

SELECT create_distributed_table('simple', 'id');
#  create_distributed_table 
# --------------------------
#  
# (1 row)

PREPARE simple_realtime AS SELECT *, $1 FROM simple;
# PREPARE

PREPARE simple_router AS SELECT *, $1 FROM simple WHERE id = 1;
# PREPARE

EXPLAIN EXECUTE simple_realtime(1);
#                                      QUERY PLAN
# -------------------------------------------------------------------------------------
#  Custom Scan (Citus Real-Time)  (cost=0.00..0.00 rows=0 width=0)
#    Task Count: 32
#    Tasks Shown: One of 32
#    ->  Task
#          Node: host=citus2009_worker_1 port=5432 dbname=postgres
#          ->  Seq Scan on simple_102040 simple  (cost=0.00..22.70 rows=1270 width=68)
# (6 rows)

EXPLAIN EXECUTE simple_router(1);
# WARNING:  there is no parameter $1
#                           QUERY PLAN
# --------------------------------------------------------------
#  Custom Scan (Citus Router)  (cost=0.00..0.00 rows=0 width=0)
#    Task Count: 1
#    Tasks Shown: All
#    ->  Task
#          Error: Could not get remote plan.
# (5 rows)
```

After checking out the worker logs, shut down the cluster…

```sh
docker logs --tail=2 citus2009_worker_1
# 2018-03-14 22:37:10.808 UTC [1485] ERROR:  there is no parameter $1 at character 111
# 2018-03-14 22:37:10.808 UTC [1485] STATEMENT:  EXPLAIN (ANALYZE FALSE, VERBOSE FALSE, COSTS TRUE, BUFFERS FALSE, TIMING FALSE, FORMAT TEXT) SELECT id, name, $1 FROM public.simple_102041 simple WHERE (id = 1)

docker-compose down -v
# Removing citus2009_worker_1 ... done
# Removing citus_manager      ... done
# Removing citus_master       ... done
# Removing network citus2009_default
```

# Expected Behavior

The remote explain plans for these two queries should be relatively similar.

# Actual Behavior

As can be seen in the worker logs, `$1` is passed to the worker without any parameter value, causing the remote explain to fail.

**Comment by ozgune** [comment]
@yglh-gq -- Thank you for reporting this issue. We're going to look at this issue and provide you with an update in 2-3 days.

**Comment by ozgune** [comment]
@jasonmp85 -- Can we reproduce this issue on our end?

**Comment by jasonmp85** [comment]
@ozgune yes, though this repro case is _very_ complex in terms of what features it's using. I spent some time trying to narrow down exactly which feature is causing it to get a minimal test case and will paste that after warming up some food 🍴.

**Comment by jasonmp85** [comment]
Pasting the original issue text below because I'm going to replace it with a more focused/minimal test case…

> # Auto_explain failure "Error: Could not get remote plan."
> 
> @yglh-gq opened this issue <relative-time datetime="2018-02-14T06:05:03Z" title="Feb 13, 2018, 11:05 PM MST">23 days ago</relative-time>
>
> ## TEST ENV
> PG9.6.2 with citus 7.1
> 
> ## PG_CONF
> We add the `auto_explain` extension in the cn node.Here is my postgresql.conf.
> 
> ```
> shared_preload_libraries = 'citus,auto_explain,...'
> auto_explain.log_min_duration = 5s
> auto_explain.log_nested_statements = on
> ```
> 
> ## DDL
> ```
> create table test_tb1(c1 int primary key,c2 int);
> create table test_tb2(c1 int primary key,c2 int);
> select create_reference_table('test_tb2');
> 
> CREATE OR REPLACE FUNCTION test_trigger()
> returns trigger
> as $$
> BEGIN
>     if (TG_OP = 'INSERT') then
>         EXECUTE 'INSERT INTO test_tb2 (c1,c2) VALUES ($1,$2)' USING NEW.c1,NEW.c2;
>     elsif (TG_OP = 'UPDATE') then
>         EXECUTE 'UPDATE test_tb2 SET c1 = $1,c2 = $2 WHERE c1 = $3' USING NEW.c1,NEW.c2,OLD.c1;
>     end if;
>     RETURN NULL;
> END;
> $$ language plpgsql;
> 
> create trigger sync_trigger after insert or update on test_tb1 FOR EACH ROW
> EXECUTE PROCEDURE test_trigger();
> 
> insert into test_tb1 select id,id from generate_series(1,100) id;
> ```
> ## Reproduce the problem
> use the `insert update` sql with the primary key conflict.
> make the execute time of the update sql long enough to trigger auto_explain,for example make a lock wait event.
> 
> worker node:
> ```
> begin;
> truncate test_tb2_102060 ;
> ```
> cn node:
> ```
> insert into test_tb1 values (6,999) on conflict(c1) do update set c1 = excluded.c1,c2 = excluded.c2;
> ```
> ## ERROR MESSAGES
> ```
> WARNING:  there is no parameter $3
> WARNING:  there is no parameter $3
> WARNING:  prepared transaction with identifier "citus_0_1584_11_30" does not exist
> WARNING:  failed to commit transaction on localhost:9701
> WARNING:  failed to roll back prepared transaction 'citus_0_1584_11_30'
> HINT:  Run "COMMIT PREPARED 'citus_0_1584_11_30'" on localhost:9701
> WARNING:  prepared transaction with identifier "citus_0_1584_11_31" does not exist
> WARNING:  failed to commit transaction on localhost:9702
> WARNING:  failed to roll back prepared transaction 'citus_0_1584_11_31'
> HINT:  Run "COMMIT PREPARED 'citus_0_1584_11_31'" on localhost:9702
> WARNING:  prepared transaction with identifier "citus_0_1584_11_32" does not exist
> 
> psql LOG:  duration: 50702.405 ms  plan:
>         Query Text: UPDATE test_tb2 SET c1 = $1,c2 = $2 WHERE c1 = $3
>         Custom Scan (Citus Router)  (cost=0.00..0.00 rows=0 width=0)
>           Task Count: 1
>           Tasks Shown: All
>           ->  Task
>                 Error: Could not get remote plan.
> ```
> and data in two tables are different now!
> ```
> test_citus=# select * from test_tb1 where c1 = 6;
>  c1 | c2  
> ----+-----
>   6 | 999
> (1 row)
> 
> test_citus=# select * from test_tb2 where c1 = 6;
>  c1 | c2 
> ----+----
>   6 |  6
> (1 row)
> ```

**Comment by jasonmp85** [comment]
Oh, and @yglh-gq: I almost forgot. We definitely want to fix the underlying issue, but at least in some cases there is a workaround…

If you take all lines in your trigger of the form `EXECUTE '<SQL WITH PARAMS>' USING foo, bar, baz`, and replace them with `EXECUTE format('<SQL WITH PARAMS>', foo, bar, baz)`, then the SQL string will be generated on the coordinator and the worker will not throw an error. Be sure to use the proper format specifiers (see `%I` for identifier and `%L` for literal values in particular), but otherwise that should fix your problem for now.

That said, it's only a workaround and I think this bug should be addressed.

**Comment by jasonmp85** [comment]
I poked a little more into this and it turns out the underlying cause is simply that the planner being used is router. The reason `dist_func` works is because it's using the real-time planner. In my example if you throw e.g. an `id = 1` onto the `SELECT` query within `dist_func`, it also fails. It just happens that reference tables trigger router execution in this case.

The simplest repro case is now:

  1. Create a distributed table
  2. Create a router-executable parameterized prepared statement (where the parameterization isn't even related to the distribution column)
  3. Try to `EXPLAIN EXECUTE` that statement

**Comment by yglh-gq** [comment]
@jasonmp85 Okay, got it. I turn off the `auto_explain` for now to avoid this problem. And is there any idea why the router executor will cause this problem?

**Comment by jasonmp85** [comment]
@yglh-gq not yet; I've reproduced the issue on Citus 7.3 and reduced the test case to something pretty simple (no `auto-explain`, even). We'll scope the fix soon and can get back to you when we have an estimate.

**Comment by marcocitus** [comment]
There's still a problem that is specific to `EXPLAIN EXECUTE`, which isn't easy to solve, but a few other prepared statement issues (without EXPLAIN) have been fixed in recent releases.

**Comment by onderkalaci** [comment]
I guess this is similar:

```SQL
PREPARE p3 (int, int) AS insert into users_table ( value_1, user_id) select value_1, user_id + $2  FROM users_table ON CONFLICT (user_id) DO UPDATE SET value_2 = EXCLUDED.value_1 + $1;

 explain analyze execute p3(20,20);
ERROR:  there is no parameter $1
CONTEXT:  while executing command on localhost:9700
```

**Comment by metdos** [comment]
Hi @pykello, is this issue fixed with #4068?

**Comment by metdos** [comment]
Fixed by #4068
